File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        ans = []
        ans.append(nums[0])
        for x in nums:
            if x > ans[-1]:
                ans.append(x)
            else:
                ind = bisect_left(ans, x)
                ans[ind] = x

        return len(ans)

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int>ans;
        ans.push_back(nums[0]);
        for(auto i : nums){
            if(i > ans[ans.size()-1]){
                ans.push_back(i);
            }
            else{
                auto it = lower_bound(ans.begin(), ans.end(), i);
                *it = i;
            }
        }
        return ans.size();
    }
};